drivers/wifi/nrf_wifi/src/wifi_debug.c

```diff
@@ -0,0 +1,372 @@
+	reg_type = strtoul(argv[1], &ptr, 10);
+	reg_offset = strtoul(argv[2], &ptr, 10);
+	val = strtoul(argv[3], &ptr, 10);
+
+	if (reg_type == 0) {
+		reg_start_addr = RPU_ADDR_SBUS_START + (reg_offset * 4);
+		reg_end_addr = RPU_ADDR_SBUS_END;
+	} else if (reg_type == 1) {
+		reg_start_addr = RPU_ADDR_PBUS_START + (reg_offset * 4);
+		reg_end_addr = RPU_ADDR_PBUS_END;
+	} else {
+		shell_fprintf(sh,
+			      SHELL_ERROR,
+			      "Invalid register type(%lu).\n",
+			      reg_type);
+		return -ENOEXEC;
+	}
+
+	if ((reg_start_addr % 4) != 0) {
+		shell_fprintf(sh,
+			      SHELL_ERROR,
+			      "Invalid register offset(%lu). Needs to be a multiple of 4\n",
+			      reg_offset);
+		return -ENOEXEC;
+	}
+
+	if (reg_start_addr + 3 > reg_end_addr) {
+		shell_fprintf(sh,
+			      SHELL_ERROR,
+			      "Invalid register offset. Exceeds bus region\n");
+		return -ENOEXEC;
+	}
+
+	status = hal_rpu_reg_write(fmac_dev_ctx->hal_dev_ctx,
+				   reg_start_addr,
+				   val);
+
+	if (status != NRF_WIFI_STATUS_SUCCESS) {
+		shell_fprintf(sh,
+			      SHELL_ERROR,
+			      "Failed to write register at 0x%x.\n",
+			      reg_start_addr);
+		return -ENOEXEC;
+	}
+
+	return 0;
+}
+
+
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	nrf_wifi_dbg_subcmds,
+	SHELL_CMD_ARG(read_mem,
+		      NULL,
+		      "<mem_type> <offset> <num_words>\n"
+		      "where:\n"
+		      "<mem_type> : One of the memory regions below\n"
+		      "0 - PKTRAM\n"
+		      "1 - GRAM\n"
+		      "<offset> : Word offset in the memory region\n"
+		      "<num_words> : Number of words to read\n",
+		      nrf_wifi_dbg_read_mem,
+		      4,
+		      0),
+	SHELL_CMD_ARG(write_mem,
+		      NULL,
+		      "<mem_type> <offset> <val>\n"
+		      "where:\n"
+		      "<mem_type> : One of the memory regions below\n"
+		      "0 - PKTRAM\n"
+		      "1 - GRAM\n"
+		      "<offset> : Word offset in the memory region\n"
+		      "<val> : Value to be written\n",
+		      nrf_wifi_dbg_write_mem,
+		      4,
+		      0),
+	SHELL_CMD_ARG(read_reg,
+		      NULL,
+		      "<reg_type> <offset> <num_regs>\n"
+		      "where:\n"
+		      "<reg_type> : One of the bus regions below\n"
+		      "0 - SYSBUS\n"
+		      "1 - PBUS\n"
+		      "<offset> : Register offset\n"
+		      "<num_words> : Number of registers to read\n",
+		      nrf_wifi_dbg_read_reg,
+		      4,
+		      0),
+	SHELL_CMD_ARG(write_reg,
+		      NULL,
+		      "<reg_type> <offset> <val>\n"
+		      "where:\n"
+		      "<reg_type> : One of the bus regions below\n"
+		      "0 - SYSBUS\n"
+		      "1 - PBUS\n"
+		      "<offset> : Register offset\n"
+		      "<val> : Value to be written\n",
+		      nrf_wifi_dbg_write_reg,
+		      4,
+		      0),
+	SHELL_SUBCMD_SET_END
+);
+
+
+SHELL_CMD_REGISTER(wifi_dbg,
+		   &nrf_wifi_dbg_subcmds,
+		   "nRF Wi-Fi debug shell commands",
+		   NULL);
+
+
+static int nrf_wifi_dbg_init(void)
+{
+	return 0;
+}
+
+
+SYS_INIT(nrf_wifi_dbg_init,
+	 APPLICATION,
+	 CONFIG_APPLICATION_INIT_PRIORITY);
```

Comment:
I do not get why this is needed, the init function is not doing anything so you can remove it.